src/decorators/apiDecoratorFactory.ts

```diff
@@ -0,0 +1,56 @@
+import {
+  applyDecorators,
+  Delete,
+  Get,
+  Post,
+  Put,
+  UseGuards,
+} from '@nestjs/common';
+import { ApiBearerAuth, ApiOkResponse, ApiOperation } from '@nestjs/swagger';
+
+import { Roles } from '@/auth/roles.decorator';
+import { RestAuthorizationGuard } from '@/graphql/rest-authorization.guard';
+import { Role } from '@/util/constants';
+
+interface ApiAuthOptions {
+  method: 'GET' | 'POST' | 'PUT' | 'DELETE';
+  path: string;
+  summary: string;
+  description: string;
+  authRoute: boolean;
+  responseType: any;
+  roles?: Role;
+}
+
+export function ApiAuthOperation(options: ApiAuthOptions) {
+  const decorators = [
+    ApiOperation({
+      summary: options.summary,
+      description: options.description,
+    }),
+    ApiOkResponse({ description: options.summary, type: options.responseType }),
+    options.authRoute && ApiBearerAuth('access-token'),
+    options.authRoute &&
+      options.roles &&
+      options.roles.length &&
+      Roles(options.roles) &&
+      UseGuards(RestAuthorizationGuard),
```

Comment:
In this code, it’s confusing when you’re adding a new condition (AND) versus when you’re using the Nullish Coalescing Operator. It would be clearer with some parentheses:
(options.authRoute && options.roles && options.roles.length) &&
(Roles(options.roles) && UseGuards(RestAuthorizationGuard))
I’m not sure if this is the logic you intended to implement, but I think this part could be made clearer.

Reply:
```
Decorator composition[#](https://docs.nestjs.com/custom-decorators#decorator-composition)
Nest provides a helper method to compose multiple decorators. For example, suppose you want to combine all decorators related to authentication into a single decorator. This could be done with the following construction:

auth.decorator.tsJS

import { applyDecorators } from '@nestjs/common';

export function Auth(...roles: Role[]) {
  return applyDecorators(
    SetMetadata('roles', roles),
    UseGuards(AuthGuard, RolesGuard),
    ApiBearerAuth(),
    ApiUnauthorizedResponse({ description: 'Unauthorized' }),
  );
}
```

Reply:
Changed the conditions to apply UseGuards and Roles Decorator;

using the  return applyDecorators(...decorators); 
at the end of ApiAuthOperation to keep conditionally addition of other decorators: (Get, Put, Delete)